frontend/src/AppState/AppState.tsx

```diff
@@ -19,56 +20,54 @@ export type UserStats = {
 };
 
 type AppStateContextState = {
-    userInfo?: UserInfo;
     appState: AppState;
     api: APIHelper;
-    site?: SiteWithUserInfo;
-    userStats: UserStats;
 };
 
 export type AppStateSetters = {
-    setSite: Dispatch<SetStateAction<SiteWithUserInfo | undefined>>;
+    site: SiteWithUserInfo | undefined;
+    userStats: UserStats;
+
+    setLoading: (value: boolean) => void;
+    setSite: (site: SiteWithUserInfo | undefined) => void;
     setUserInfo: (user: UserInfo | undefined) => void;
-    setAppState: (state: AppState) => void;
-    setUserStats: Dispatch<SetStateAction<UserStats>>;
+    setUserStats: (stats: UserStats) => void;
 }
 
 const AppStateContext = createContext<AppStateContextState>({} as AppStateContextState);
 
 const apiBase = new APIBase();
 
 export const AppStateProvider = (props: {children: ReactNode}) => {
-    const [userInfo, setUserInfo] = useState<UserInfo | undefined>(undefined);
-    const [appState, setAppState] = useState<AppState>(AppState.loading);
-    const [userStats, setUserStats] = useState<UserStats>({ notifications: 0, watch: { posts: 0, comments: 0 } });
-    const [site, setSite] = useState<SiteInfo>();
-    const api = useMemo(() => new APIHelper(apiBase, {setUserInfo, setAppState, setSite, setUserStats}), []);
-    api.updateSetters({setUserInfo, setAppState, setSite, setUserStats});
+    const appState = useMemo(() => new AppState(), []);
+    const api = useMemo(() => new APIHelper(apiBase, appState), []);
 
     useEffect(() => {
         api.init().then();
     }, [api]);
 
     useEffect(() => {
-        const link = document.querySelector('link[rel~="icon"]') as HTMLLinkElement;
-        if (!link) {
-            return;
-        }
-        if (userStats && userStats.notifications > 0) {
-            link.href = '//' + process.env.REACT_APP_ROOT_DOMAIN + '/favicon-badge.png';
-        }
-        else {
-            link.href = '//' + process.env.REACT_APP_ROOT_DOMAIN + '/favicon.ico';
-        }
-    }, [userStats]);
-
-    return <AppStateContext.Provider value={{ userInfo, appState, api: api, site: site, userStats }}>
+        autorun(() => {
```

Comment:
No, you are right, since the code inside `autorun` doesn’t interfere with react rendering, it’s ok to use it without wrapping in `useEffect`.

Reply:
Huh. I started looking into mobx and now I think nothing wrong with `autorun` inside `useEffect` but it should return a disposer. It does not matter in this case, since it is root ContextProvider and it should never to be rerendered now, but in general there is nothing wrong (except disposer).

Reply:
Yep, you’re right.